stdlib/mexpr/symbolize.mc

```diff
@@ -128,15 +129,20 @@ lang DataSym = DataAst
 end
 
 lang MatchSym = MatchAst
-  sem symbolizePat (env : Env) =
+  -- TODO(vipa): env is constant throughout symbolizePat, so it would be preferrable to pass it
+  -- in some other way, a reader monad or something. patEnv on the other hand changes, it would
+  -- be nice to pass via state monad or something.
+  -- env is the environment from the outside, plus the names added thus far in the pattern
+  -- patEnv is only the newly added names
+  sem symbolizePat (env : Env) (patEnv : Env) =
```

Comment:
Is it strictly needed have both `env` and `patEnv` in order to implement the new patterns? I find it cleaner to just pass in an environment which gets incrementally extended throughout `symbolize`, avoiding the need to merge the environments afterwards.

Reply:
We need to introduce new symbols for names that are introduced in a pattern, but keep the same symbol if the same name is introduced multiple times. This shows up in `|` patterns, e.g., `"a" ++ rest | "b" ++ rest` should have the same symbol for both `rest` names. If we just pass along a single environment I don't think we can distinguish symbols introduced previously in the same pattern vs those introduced in a parent expression, e.g., `let a = 1 in match 1 with a then ...`.

Reply:
Ah, of course, the pattern names should introduce new symbols just as lambdas do. Then this is actually a bug in my first implementation!